connectors/source/firecrawl.py

```diff
@@ -0,0 +1,312 @@
+        # Poll until completion or timeout
+        while True:
+            # Check the crawl status
+            result = firecrawl.check_crawl_status(crawl_id)
+            
+            if result.get("status") == "completed":
+                break
+                
+            # Check for timeout
+            if time.time() - start_time > timeout:
+                return {"error": f"Timeout waiting for crawl job {crawl_id} to complete"}
+                
+            # Wait before polling again
+            await asyncio.sleep(poll_interval)
+        
+        # Create a temporary directory for HTML files
+        with tempfile.TemporaryDirectory() as temp_dir:
+            # Create a crawl-specific subdirectory to keep files organized
+            crawl_dir = os.path.join(temp_dir, crawl_id)
+            os.makedirs(crawl_dir, exist_ok=True)
+            
+            # Get the crawl results
+            scrape_result = firecrawl.check_crawl_status(crawl_id)
+            
+            # Process scrape_result['data'], which is a list of dicts, each with an 'html' key
+            file_paths = []
+            if 'data' in scrape_result and isinstance(scrape_result['data'], list):
+                for i, page_data in enumerate(scrape_result['data']):
+                    # Skip if no HTML content
+                    if 'html' not in page_data:
+                        continue
+                    
+                    # Get the URL from metadata if available, otherwise use index
+                    url = page_data.get("metadata", {}).get("url", f"page-{i}")
+                    content = page_data.get("html", f"<html><body>Content for {url}</body></html>")
+                    
+                    # Clean the URL to create a valid filename
+                    filename = url.replace("https://", "").replace("http://", "")
+                    filename = filename.replace("/", "_").replace("?", "_").replace("&", "_")
+                    filename = filename.replace(":", "_")  # Additional character cleaning
+                    
+                    # Ensure the filename isn't too long
+                    if len(filename) > 200:
+                        # Use the domain and a hash of the full URL if too long
+                        domain = filename.split('_')[0]
+                        filename_hash = hashlib.md5(url.encode()).hexdigest()
+                        filename = f"{domain}_{filename_hash}.html"
+                    else:
+                        filename = f"{filename}.html"
+                    
+                    file_path = os.path.join(crawl_dir, filename)
+                    
+                    # Write the HTML content to file
+                    with open(file_path, "w", encoding="utf-8") as f:
+                        f.write(content)
+                    file_paths.append(file_path)
+            
+            # Construct the final S3 URI with crawl_id directly
+            final_s3_uri = f"{s3_uri}{crawl_id}/"
```

Comment:
Should there be anything between {s3_uri} and {crawl_id}? Or check if anything might be required?